pydicom/cli/main.py

```diff
@@ -0,0 +1,219 @@
+
+    # If list empty after pop above, then have pydicom::filename
+    if not prefix_file:
+        return prefix, last, ""
+
+    return prefix, "".join(prefix_file), last
+
+
+def filespec_parser(filespec: str):
+    """Utility to return a dataset and an optional data element value within it
+
+    Note: this is used as an argparse 'type' for adding parsing arguments.
+
+    Parameters
+    ----------
+    filespec: str
+        A filename with optional `pydicom::` prefix and optional data element,
+        in format:
+            [pydicom::]<filename>[::<element>]
+        If an element is specified, it must be a path to a data element,
+        sequence item (dataset), or a sequence.
+        Examples:
+            your_file.dcm
+            your_file.dcm::StudyDate
+            pydicom::rtplan.dcm::BeamSequence[0]
+            pydicom::rtplan.dcm::BeamSequence[0].BeamLimitingDeviceSequence
+
+    Returns
+    -------
+    matches: List[Tuple[Dataset, Any]]
+        Matching pairs of (dataset, data element value)
+        This usually is a single pair, but a list is returned for future
+        ability to work across multiple files.
+
+    Note
+    ----
+        This function is meant to be used in a call to an `argparse` libary's
+        `add_argument` call for subparsers, with name="filespec" and
+        `type=filespec_parser`. When used that way, the resulting args.filespec
+        will contain the return values of this function
+        (e.g. use `ds, element_val = args.filespec` after parsing arguments)
+        See the `pydicom.cli.show` module for an example.
+
+    Raises
+    ------
+    argparse.ArgumentTypeError
+        If the filename does not exist in local path or in pydicom test files,
+        or if the optional element is not a valid expression,
+        or if the optional element is a valid expression but does not exist
+        within the dataset
+    """
+    prefix, filename, element = filespec_parts(filespec)
+
+    # Get the pydicom test filename even without prefix, in case user forgot it
+    try:
+        pydicom_filename = get_testdata_file(filename)
+    except NotImplementedError:  # will get this if absolute path passed
+        pydicom_filename = ""
+
+    if prefix == "pydicom":
+        filename = pydicom_filename
+
+    # Check element syntax first to avoid unnecessary load of file
+    if element and not re_file_spec_object.match(element):
+        raise argparse.ArgumentTypeError(
+            f"Component '{element}' is not valid syntax for a "
+            "data element, sequence, or sequence item"
+        )
+
+    # Read DICOM file
+    try:
+        ds = dcmread(filename, force=True)
+    except FileNotFoundError:
+        extra = (
+            (f", \nbut 'pydicom::{filename}' test data file is available")
+            if pydicom_filename
+            else ""
+        )
+        raise argparse.ArgumentTypeError(f"File '{filename}' not found{extra}")
+    except Exception as e:
```

Comment:
Maybe check for `InvalidDicomError` before that with an appropriate error message

Reply:
I'm using `force=True` so not much chance of InvalidDicomError, and I'm not clear on what useful info I could add to the message before re-raising it.

Reply:
Ah, you are right - I mixed that up. I actually had a quick look at `dmread` and found
```
    Raises
    ------
    InvalidDicomError
        If `force` is ``True`` and the file is not a valid DICOM file.
```
which is wrong, I think.